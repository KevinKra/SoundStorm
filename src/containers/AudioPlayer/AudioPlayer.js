import React, { Component } from "react";
import { connect } from "react-redux";
import * as actions from "../../_redux/actions";
import "./AudioPlayer.scss";

class AudioPlayer extends Component {
  constructor(props) {
    super(props);
    this.audio = React.createRef();
  }

  playSong = () => {
    this.props.playSong(this.props.currentSong);
    this.audio.current.play();
  };

  pauseSong = () => {
    this.props.pauseSong(this.props.currentSong);
    this.audio.current.pause();
  };

  nextSong = i => {
    const newSong = this.props.currentPlaylist[i + 1];
    const oldSong = this.props.currentPlaylist[i];
    if (newSong !== undefined) {
      oldSong.playing = false;
      newSong.playing = true;
      this.props.incrementIndex(
        this.props.currentSongIndex,
        this.props.currentPlaylist.length - 1
      );
      this.props.playTargetSong(newSong);
    }
    if (newSong === undefined) {
      const beginning = this.props.currentPlaylist[0];
      oldSong.playing = false;
      beginning.playing = true;
      this.props.incrementIndex(
        this.props.currentSongIndex,
        this.props.currentPlaylist.length - 1
      );
      this.props.playTargetSong(beginning);
    }
  };

  prevSong = i => {
    const newSong = this.props.currentPlaylist[i - 1];
    const oldSong = this.props.currentPlaylist[i];
    if (newSong !== undefined) {
      oldSong.playing = false;
      newSong.playing = true;
      this.props.decrementIndex(
        this.props.currentSongIndex,
        this.props.currentPlaylist.length - 1
      );
      this.props.playTargetSong(newSong);
    }
    if (newSong === undefined) {
      const end = this.props.currentPlaylist[
        this.props.currentPlaylist.length - 1
      ];
      oldSong.playing = false;
      end.playing = true;
      this.props.decrementIndex(
        this.props.currentSongIndex,
        this.props.currentPlaylist.length - 1
      );
      this.props.playTargetSong(end);
    }
  };

  render() {
    const handlePlayIcon = this.props.currentSong.playing ? (
      <i className="fas fa-lg fa-pause-circle" onClick={this.pauseSong} />
    ) : (
      <i className="fas fa-lg fa-play-circle" onClick={this.playSong} />
    );

    return (
      <section className="AudioPlayer">
        <section className="song-info">
          <img
            src={this.props.currentSong.album_image}
            alt={this.props.currentSong.album_name}
          />
          <div className="artist-album-info">
            <p>{this.props.currentSong.artist_name}</p>
            <p>{this.props.currentSong.name}</p>
          </div>
        </section>
        <audio
          src={this.props.currentSong.audio}
          autoPlay={true}
          ref={this.audio}
          onEnded={() => this.nextSong(this.props.currentSongIndex)}
        />
        <div className="button-panel">
          <i
            className="fas fa-lg fa-step-backward"
            onClick={() => this.prevSong(this.props.currentSongIndex)}
          />
          {handlePlayIcon}
          <i
            className="fas fa-lg fa-step-forward"
            onClick={() => this.nextSong(this.props.currentSongIndex)}
          />
        </div>
      </section>
    );
  }
}

const mapStateToProps = store => ({
  currentSong: store.currentSong,
  currentPlaylist: store.currentPlaylist,
  currentSongIndex: store.currentSongIndex
});

const mapDispatchToProps = dispatch => ({
  playSong: song => dispatch(actions.playSong(song)),
  pauseSong: song => dispatch(actions.pauseSong(song)),
  playTargetSong: song => dispatch(actions.playTargetSong(song)),
  incrementIndex: (currentIndex, playlistLength) =>
    dispatch(actions.incrementIndex(currentIndex, playlistLength)),
  decrementIndex: (currentIndex, playlistLength) =>
    dispatch(actions.decrementIndex(currentIndex, playlistLength))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AudioPlayer);
