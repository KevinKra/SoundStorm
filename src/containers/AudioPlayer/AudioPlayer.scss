.AudioPlayer {
  display: grid;
  grid-template-columns: 300px 1fr 1fr;
  grid-template-rows: 8vh;
  // border: 1px dashed red;
  .song-info {
    display: flex;
    // border: 1px solid orange;
  }
  .button-panel {
    // border: 1px solid blue;
    display: flex;
    align-items: center;
    justify-content: space-around;
  }
  .artist-album-info {
    // border: 1px solid red;
    padding-left: 0.5rem;
    // width: 100%;
  }
  h4,
  p {
    margin: 0;
  }
  i {
    // margin: 0 50px;
  }
  img {
    height: 8vh;
  }
}
