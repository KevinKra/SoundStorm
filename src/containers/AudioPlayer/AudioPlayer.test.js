import React from "react";
import { AudioPlayer } from "./AudioPlayer";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const mockCurrentSong = {
  id: "1197872",
  name: "Perverting The American Dream",
  duration: 238,
  artist_id: "462184",
  artist_name: "The Bourgeois",
  artist_idstr: "The_Bourgeois",
  album_name: "The Bourgeois",
  album_id: "144113",
  position: 1,
  releasedate: "2015-02-16",
  album_image:
    "https://imgjam2.jamendo.com/albums/s144/144113/covers/1.200.jpg",
  audio:
    "https://mp3l.jamendo.com/?trackid=1197872&format=mp31&from=app-d011137e",
  audiodownload: "https://mp3d.jamendo.com/download/track/1197872/mp32/",
  shorturl: "http://jamen.do/t/1197872",
  shareurl: "http://www.jamendo.com/track/1197872",
  image: "https://imgjam2.jamendo.com/albums/s144/144113/covers/1.200.jpg",
  playing: true
};

describe("<AudioPlayer", () => {
  let wrapper;
  const props = {
    pauseSong: jest.fn(),
    playSong: jest.fn(),
    incrementIndex: jest.fn(),
    decrementIndex: jest.fn()
  };
  beforeEach(() => {
    wrapper = shallow(<AudioPlayer currentSong={mockCurrentSong} {...props} />);
  });
  it("should match the snapshot", () => {
    console.log(wrapper.debug());
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  xit("should pause a song when the pause button is pressed", () => {
    wrapper.find("i.fa-pause-circle").simulate("click");
  });
  // it("should play a song when the play button is pressed", () => {
  //   wrapper.find("i.fa-play-circle").simulate("click");
  // });
  // it("should iterate to a new song when nextSong is invoked", () => {
  //   wrapper.find("i.fa-step-forward").simulate("click");
  // });
  // it("should iterate to a previous song when prevSong is invoked", () => {
  //   wrapper.find("i.fa-step-backward").simulate("click");
  // });
});
