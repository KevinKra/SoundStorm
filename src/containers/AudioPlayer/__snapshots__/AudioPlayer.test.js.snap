// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioPlayer should match the snapshot 1`] = `
<section
  className="AudioPlayer"
>
  <section
    className="song-info"
  >
    <img
      alt="The Bourgeois"
      src="https://imgjam2.jamendo.com/albums/s144/144113/covers/1.200.jpg"
    />
    <div
      className="artist-album-info"
    >
      <p>
        The Bourgeois
      </p>
      <p>
        Perverting The American Dream
      </p>
    </div>
  </section>
  <audio
    autoPlay={true}
    onEnded={[Function]}
    src="https://mp3l.jamendo.com/?trackid=1197872&format=mp31&from=app-d011137e"
  />
  <div
    className="button-panel"
  >
    <i
      className="fas fa-lg fa-step-backward"
      onClick={[Function]}
    />
    <i
      className="fas fa-lg fa-pause-circle"
      onClick={[Function]}
    />
    <i
      className="fas fa-lg fa-step-forward"
      onClick={[Function]}
    />
  </div>
</section>
`;
